File: haskell/nucleotide-count/DNA.hs
module DNA (count, nucleotideCounts) where

import Data.List as L
import Data.Map as M

count :: Char -> String -> Either String Int
count char dna = do
  c <- toNucleotide char
  counts <- nucleotideCounts dna
  return $ counts ! c

nucleotideCounts :: String -> Either String (Map Char Int)
nucleotideCounts = foldDNA countNucleotides zeroCounts
  where
    countNucleotides = flip $ M.adjust succ
    zeroCounts = M.fromList (L.map zeroCount nucleotides)
    zeroCount nucl = (nucl, 0)

foldDNA :: (a -> Char -> a) -> a -> String -> Either String a
foldDNA f zero = L.foldl accumulate (Right zero)
  where
    accumulate a char = do
      acc <- a
      nucleotide <- toNucleotide char
      return $ f acc nucleotide

toNucleotide :: Char -> Either String Char
toNucleotide char =
  if char `elem` nucleotides
    then Right char
    else Left $ "invalid nucleotide '" ++ [char] ++ "'"

nucleotides :: String
nucleotides = "ACGT"

File: scala/nucleotide-count/src/main/scala/DNA.scala
import DNA._

class DNA(dna: String) {

  require(isValidDNA(dna),
      s"DNA must contain only nucleotides ${DnaNucleotides.mkString}: $dna")

  lazy val nucleotideCounts: NucleotideCounts = {
    val counts = dna groupBy(identity) mapValues(_.length)
    EmptyDnaNucleotideCounts ++ counts
  }
}

object DNA {
  type Nucleotide = Char
  type NucleotideCounts = Map[Nucleotide,Int]

  val DnaNucleotides: Set[Nucleotide] = Set('A', 'T', 'C', 'G')

  val EmptyDnaNucleotideCounts: NucleotideCounts = {
    def zeroCount(nucleotide: Nucleotide) = nucleotide -> 0

    DnaNucleotides map zeroCount toMap
  }

  def isValidDNA(dna: String) =
    dna forall DnaNucleotides.contains
}